modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -33,100 +36,164 @@ trait ConstructionService[F[_]] {
+                .map(pk => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError])
+            }.flatMap { proof =>
+              EitherT
+                .fromEither(
+                  KryoSerializer[F]
+                    .serialize(Signed[DAGTransaction](transaction, NonEmptySet.of(proof)))
+                )
+                .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
+                .map(Hex.fromBytes(_))
+            }
+          }
+
+      override def getPayloads(
+        operations: NonEmptyList[Operation],
+        metadata: TransactionMetadata
+      ): EitherT[F, ConstructionError, PayloadsResult] =
+        for {
+          _ <- EitherT.cond(operations.size == 2, (), IncorrectNumberOfOperations)
+
+          (positiveOperation, negativeOperation) <- EitherT.fromEither {
+            val first = operations.head
+            val second = operations.tail.head
+            if (first.amount.value.isPositive && second.amount.value.isNegative)
+              (first, second).asRight
+            else if (first.amount.value.isNegative && second.amount.value.isPositive)
+              (second, first).asRight
+            else
+              MismatchedOperations.asLeft
+          }
+
+          _ <- EitherT.cond(
+            positiveOperation.amount.negate.map(_ == negativeOperation.amount).getOrElse(false),
+            (),
+            MismatchedOperations
+          )
+
+          transactionAmount <- positiveOperation.amount.value.toTransactionAmount
+            .leftMap(_ => MalformedTransaction) // This should never happen
+            .toEitherT
+
+          sourceAddress = negativeOperation.account.address
+          transactionFee <- EitherT.fromEither(
+            metadata.suggestedFee.headOption match {
+              case None      => TransactionFee.zero.asRight
+              case Some(amt) => amt.value.toTransactionFee.leftMap(_ => InvalidSuggestedFee)
+            }
+          )
+
+          salt <- EitherT.right(saltGen())
+          dagTransaction =
+            DAGTransaction(
+              source = sourceAddress,
+              destination = positiveOperation.account.address,
+              amount = transactionAmount,
+              fee = transactionFee,
+              parent = metadata.lastReference,
+              salt = salt
+            )
+
+          serializedTxn <-
+            KryoSerializer[F]
+              .serialize(dagTransaction)
+              .leftMap(e => SerializationError(e.getMessage))
+              .toEitherT
+
+          signBytes <-
+            serializedTxn.hash
+              .leftMap[ConstructionError](_ => MalformedTransaction)
+              .map(h => Hex(h.value))
+              .toEitherT
+
+          payload = SigningPayload(AccountIdentifier(sourceAddress, none), signBytes, SignatureType.ECDSA)
+
+        } yield PayloadsResult(Hex.fromBytes(serializedTxn), NonEmptyList.one(payload))
```

Comment:
I see that you use `EitherT` mostly as `toEitherT` or `EitherT.cond`, just because the returned type is `EitherT`. I'd suggest to delay lifting it to `EitherT to the very end, if you actually don't need that before

Reply:
The one exception is the generation of the salt, which evaluates to `F[TransactionSalt]`

Reply:
yep, so it can stay in the effect, map or flatMap and at the end you can lift to monad transformer